### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by derekfairholm**
Completed Part 1

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -56,64 +56,78 @@ Create a branch named Part1
...
-    T(<#type name#> v, const char* <#variable name#>)   //1
-    //2
-    //3
+    int value;
+    std::string name;
+    
+    T(int v, const char* name)    //1 . I'm wondering about casting from char* to std::string here? It only shows the firs character in the console
+    {
+        value = v;    //2
+        this->name = *name;    //3
```
Comment (matkatmusic) [inline main.cpp]: std::string can be initialized with `const char*`.  don't dereference or initialize in-class.  use member initializer list to init it.  Also, change your arg name to not overshadow the member variable.

- [x] rename arg

- [x] init members via member initializer list.

File: main.cpp [hunk 2]
```diff
@@ -56,64 +56,78 @@ Create a branch named Part1
...
 {
-    <#type name#> compare(<#type name#> a, <#type name#> b) //5
+    T* compare(T* a, T* b)    //5
     {
-        if( a->value < b->value ) return a;
-        if( a->value > b->value ) return b;
+        if(a->value < b->value) return a;
+        if(a->value > b->value) return b;
         return nullptr;
     }
```
Comment (matkatmusic) [inline main.cpp]: What do we always do before using pointers?

- [x] apply safe pointer usage.

File: main.cpp [hunk 3]
```diff
@@ -56,64 +56,78 @@ Create a branch named Part1
...
 struct U
 {
-    float <#name1#> { 0 }, <#name2#> { 0 };
-    <#returnType#> <#memberFunction#>(<#type name#>* <#updatedValue#>)      //12
+    float floatOne { 0 }, floatTwo { 0 };
+    float memberFunction(float* updatedValue)    //12
     {
-        
+        std::cout << "U's floatOne value: " << floatOne << std::endl;
+        floatOne = *updatedValue;
```
Comment (matkatmusic) [inline main.cpp]: What do we always do before using pointers?

- [x] apply safe pointer usage.

File: main.cpp [hunk 4]
```diff
@@ -56,64 +56,78 @@ Create a branch named Part1
...
 {
-    static <#returntype#> <#staticFunctionA#>(U* that, <#type name#>* <#updatedValue#> )        //10
+    static float staticFunctionA(U* that, float* updatedValue)    //10
     {
-        std::cout << "U's <#name1#> value: " << that-><#name1#> << std::endl;
-        that-><#name1#> = <#updatedValue#>;
-        std::cout << "U's <#name1#> updated value: " << that-><#name1#> << std::endl;
-        while( std::abs(that-><#name2#> - that-><#name1#>) > 0.001f )
+        std::cout << "U's floatOne value: " << that->floatOne << std::endl;
+        that->floatOne = *updatedValue;
```
Comment (matkatmusic) [inline main.cpp]: What do we always do before using pointers?

- [x] apply safe pointer usage.

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 1 above)
Comment (derekfairholm) [reply main.cpp]: Got it, made these changes

File: main.cpp (hunk 2 above)
Comment (derekfairholm) [reply main.cpp]: Sorry, you were really clear about safe pointer usage.

File: main.cpp [hunk 5]
```diff
@@ -56,64 +56,83 @@ Create a branch named Part1
...
+            std::cout << "U's floatOne value: " << floatOne << std::endl;
+            floatOne = *updatedValue;
+            std::cout << "U's floatOne updated value: " << floatOne << std::endl;
+            while(std::abs(floatTwo - floatOne) > 0.001f)
+            {
+                floatTwo += 0.001f;
+            }   
+            std::cout << "U's floatTwo updated value: " << floatTwo << std::endl;
+            return floatTwo * floatOne;
+        }
```
Comment (matkatmusic) [inline main.cpp:95]: add missing return statement.  just return 0. 

- [x] add missing return statement if updatedValue is nullptr

File: main.cpp [hunk 6]
```diff
@@ -56,64 +56,83 @@ Create a branch named Part1
...
+            std::cout << "U's floatOne value: " << that->floatOne << std::endl;
+            that->floatOne = *updatedValue;
+            std::cout << "U's floatOne updated value: " << that->floatOne << std::endl;
+            while( std::abs(that->floatTwo - that->floatOne) > 0.001f)
+            {
+                that->floatTwo += 0.001f;
+            }
+            std::cout << "U's floatTwo updated value: " << that->floatTwo << std::endl;
+            return that->floatTwo * that->floatOne;
         }
```
Comment (matkatmusic) [inline main.cpp]: - [x] add missing return statement

File: main.cpp [hunk 7]
```diff
@@ -56,64 +56,83 @@ Create a branch named Part1
...
+struct StructTwo
 {
-    static <#returntype#> <#staticFunctionA#>(U* that, <#type name#>* <#updatedValue#> )        //10
+    static float staticFunctionA(U* that, float* updatedValue)    //10
     {
-        std::cout << "U's <#name1#> value: " << that-><#name1#> << std::endl;
-        that-><#name1#> = <#updatedValue#>;
-        std::cout << "U's <#name1#> updated value: " << that-><#name1#> << std::endl;
-        while( std::abs(that-><#name2#> - that-><#name1#>) > 0.001f )
+        if(updatedValue != nullptr)
```
Comment (matkatmusic) [inline main.cpp]: - [x] `that` is a pointer.  so you need to check if it's null before use!

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 6 above)
Comment (derekfairholm) [reply main.cpp]: added 
`return 0;`
the function doesn't have a pointer as return value so that's the best I could do, right?

Comment (derekfairholm) [reply main.cpp]: sorry I missed your previous comment!

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**